### PR TITLE
add jute autogenerated code

### PR DESCRIPTION
**Pull request by pmazzini**
Add jute autogenerated code. This is the first step towards #28.

### CODE REVIEW

File: data/acl.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+	"github.com/go-zookeeper/jute/lib/go/jute"
+)
+
+type ACL struct {
+	Perms int32 // perms
+	Id    *Id   // id
+}
+
+func (r *ACL) GetPerms() int32 {
```
Comment (yarikk) [inline data/acl.go:17]: The generated code has lots of getters like this. Idiomatic Go avoids getters. Something to watch out for.

File: Makefile [hunk 2]
```diff
@@ -49,3 +43,14 @@ clean:
...
 	rm -f zookeeper
 	rm -f profile.cov
+
+.PHONY: jute
+jute:
+	wget $(JUTE_URL)
+	go run github.com/go-zookeeper/jute/cmd/jutec \
+		-go.moduleMap=org.apache.zookeeper: \
+		-go.prefix=github.com/go-zookeeper/zk \
+		-outDir . \
```
Comment (nemith) [inline Makefile:53]: Maybe we should put these in an internal subdirectory?

File: data/acl.go (hunk 1 above)
Comment (pmazzini) [reply data/acl.go:17]: I think that the thrift compiler does the same thing, I like the fact that it is consistent with it.

File: data/id.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
+// Autogenerated jute compiler
+// @generated from '/home/pmazzini/repos/zk/zookeeper.jute'
```
Comment (jeffbean) [inline data/id.go:2]: having the full path here is not something we want. Can we do relative paths or not include the path?

File: Makefile (hunk 2 above)
Comment (pmazzini) [reply Makefile:53]: I am not sure about it. For example the lock recipe currently takes an `ACL` type.

```
func NewLock(c *Conn, path string, acl []ACL) *Lock {
```

This ACL type is generated by jute. So we want to be able to import this package in the recipes.

File: data/id.go (hunk 3 above)
Comment (pmazzini) [reply data/id.go:2]: Yeah, I agree https://github.com/go-zookeeper/jute/issues/20

**Comment by pmazzini** [comment]
This is only adding new types, it should not be a breaking change, it is a no-op at this point.

I was planning to slowly replace some types such as https://github.com/go-zookeeper/zk/blob/master/structs.go#L31 with https://github.com/go-zookeeper/jute/blob/master/testdata/fixtures/zookeeper/data/statpersisted.go#L12 which should also be a trivial change.

I am not planning to introduce breaking changes and was looking forward to do 1:1 replacement of some structs.

**Comment by pmazzini** [comment]
This is only adding new types, it should not be a breaking change, it is a no-op at this point.

I was planning to slowly replace some types such as https://github.com/go-zookeeper/zk/blob/master/structs.go#L31 with https://github.com/go-zookeeper/jute/blob/master/testdata/fixtures/zookeeper/data/stat.go#L12 which should also be a trivial change.

I am not planning to introduce breaking changes and was looking forward to do 1:1 replacement of some structs.

**Comment by jeffbean** [comment]
> This is only adding new types, it should not be a breaking change, it is a no-op at this point.
> 
> I was planning to slowly replace some types such as https://github.com/go-zookeeper/zk/blob/master/structs.go#L31 with https://github.com/go-zookeeper/jute/blob/master/testdata/fixtures/zookeeper/data/stat.go#L12 which should also be a trivial change.
> 
> I am not planning to introduce breaking changes and was looking forward to do 1:1 replacement of some structs.

The primary concern is whether the generated code is part of the public API. The answer to that has implications to the following questions:

- Is this an API breaking change that necessitates a v2?
- What happens if this change needs to be rolled back?
- Will this be able to adapt to a change in the Zookeeper serialization format, as proposed in ZOOKEEPER-102

If the generated code is exposed as part of the public API of the library, the public types in the top level github.com/go-zookeeper/zk package will be replaced with their newly generated counterparts. This means,

1. This is an API breaking change as the generated types do not align exactly. For example, compare [ACL on master](https://godoc.org/github.com/go-zookeeper/zk#ACL) with [ACL in the generated code](https://github.com/go-zookeeper/zk/pull/31/files#diff-6191b89f9078301f0b60e97a0b5128deR12). This means a v2 is required.
2. If this change needs to be rolled back, that’s another breaking change and major version bump.
3. If the Zookeeper serialization format changes, these generated types will change again, and that is yet another breaking change and major version bump. 

Alternatively, if the generated code is not exposed as part of the public API, the existing types in the top level github.com/go-zookeeper/zk package will be retained. Internal mappers will be necessary to convert from/to public types to their internal generated variants. This means,

1. This is a completely backwards compatible change. The change in serialization is an internal detail.
2. The change may be rolled back in a safe manner in a patch release.
3. If the Zookeeper serialization format changes but the shapes are similar, the public types of go-zookeeper/zk will remain unchanged. Mappers to types generated by the new serialization format will be necessary.

A caveat of this approach is that there’s a small cost to mapping to/from custom types to generated types.

**Comment by nemith** [comment]
> This is an API breaking change as the generated types do not align exactly. For example, compare ACL on master with ACL in the generated code. This means a v2 is required.

I think exposing these structs directly was a bad choice to begin with.  This is why i suggest putting these in an `internal` package and then for the incompatible public APIs we just convert them to public types.  

Having end users fill out structs that get directly serialized seems like a bad idea .

> Will this be able to adapt to a change in the Zookeeper serialization format, as proposed in ZOOKEEPER-102

ZOOKEEPER-102 has been dead without a whole lot of work.  (see the response to the question you posted in the Google Doc.  If for some reason it get resurected then we will dump jute and move to protobufs which ALSO does codegen.  This also means seperating out the client API from the serialization is a good idea.  


> If this change needs to be rolled back, that’s another breaking change and major version bump.
We can start a development jute path and do a lot of testing, but as I stated above this can be rolled out without breaking the API



I think the answer to all your issues is that we need to separate out the API from serialization which i have tested

**Pull request closed without merging**